Update bookmarks status from main buffer

thierryvolpiatto opened the pull request:
This PR allow refreshing the main buffer with `revert-buffer` (g). Before the unread count were not updated when hitting "g", now it does.
Also I tried clarifying the call to `mu4e~start` by separating it in two functions.
Same for the `revert-buffer-function`, it is now split in two functions.
Please have a look before merging.

Next step would be to refresh the unreads count after hitting "U" and update finish automatically without having to hit "g". Same after having read the unread and quitting `mu4e-headers` .

thierryvolpiatto commented:
Update: After reading unread mails and quitting mu4e-headers, main buffer and unreads are updated correctly :-)

thierryvolpiatto commented:
Update: Counters are refreshed after updating mails and/or index.

thierryvolpiatto commented:
Dirk-Jan C. Binnema <notifications@github.com> writes:

> @djcb commented on this pull request.
>
> Thanks -- seems useful...

Yes, otherwise counters are useless, now they are updated each time the
unread status change:
1) After updating mails ("U").
2) After reading unread mails and quitting mu4e-headers.

Before this change, I had to quit mu4e and restart it to have counters
updated which is inconvenient.

The only way I found to update the main buffer was to make a call to
`mu4e~start` from the revert-buffer-function of main mode.
Problem was that `mu4e~start` calls itself `mu4e~main-view`
inconditionally (not sure it is anymore true with your last change, I
have to look).

> but I think it would be cleaner to use mu4e-index-updated-hook from
> mu4e-main.

Ok thanks, I didn't know this hook.

> Also mu4e~main-view-real-1?

It is same as the old revert-buffer function but allows calling it with
a REFRESH argument which is needed (or not) in various places
i.e. allows two behaviors of the revert-buffer function.

See also `mu4e~start` and `mu4e~start-1` and let me know what you think.

What function otherwise would you use in the `mu4e-index-updated-hook` to update
counters? (or what this function if new would invoke?)

Thanks.

--
Thierry

Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

thierryvolpiatto commented:
Here the `mu4e~start` with last changes you made:
```elisp
(defun mu4e~start (&optional func)
[...]
  ;; if we're already running, simply go to the main view
  (if (mu4e-running-p)         ;; already running?
      (when func (funcall func)) ;; yes! run func if defined
    (progn
[...]
```
With this code, calling `mu4e~start` with no arg while mu4e is running is a noop, it does nothing
perhaps what you wanted was:
```elisp
(if (and func (mu4e-running-p))
    (funcall func)
  [...])
```
Also the progn is unneeded in the else clause.

thierryvolpiatto commented:
Dirk-Jan C. Binnema <notifications@github.com> writes:

> but I think it would be cleaner to use mu4e-index-updated-hook from
> mu4e-main.

I had a look at this hook but don't understand how to use it to revert
main buffer, any help welcome.

Thanks.

PS: Otherwise this PR is working fine AFAICT as it is actually.

-- 
Thierry

Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

djcb commented:
I've updated things a bit, so you get fresh unread/all counts when calling mu4e, and added add `g` binding as well, that was indeed missing.

What's remaining is the updating bit -- that only really matters when pressing 'U' while in the main screen. I think it should be possible to use mu4e-index-updated-hook for that (or else we need to fix that!)... but anyway, I'll add that.

It might still be useful as optimization is having mu4e-index-updated-hook (and the :update / :remove messages) mark things as 'dirty'.. and skip the refresh when we're not dirty. But, at least for me, it works fine as it is.

thierryvolpiatto commented:
Dirk-Jan C. Binnema <notifications@github.com> writes:

> I've updated things a bit, so you get fresh unread/all counts when
> calling mu4e,

Thanks, but still other things wrong IMHO, see below.

> and added add g binding as well, that was indeed missing.

No, g binding was here and correctly bound to revert-buffer as we derive
from special-mode and we have set revert-buffer-function in
mu4e-main-mode.
Calling revert-buffer should refresh the buffer.

(BTW same for "q" but this is unrelated).

> What's remaining is the updating bit

Not only, quitting mu4e-headers after reading a mail doesn't update.

> -- that only really matters when pressing 'U' while in the main
> screen. I think it should be possible to use mu4e-index-updated-hook
> for that (or else we need to fix that!)... but anyway, I'll add that.

Ok, good, thanks.

One thing also which is unclear is mu4e~start, its behavior have changed
at least 2 times this week, would need to clarify what it does (or
should do) in docstring or comment for further development, I tried to
do this along your changes in this PR (see docstring) but not sure now
if it's right (i.e. my docstring):

1) When mu4e is running and FUNC is defined:
master: (re)start mu4e and call FUNC in the filter process.
before: Call FUNC only.

2) When mu4e is running and FUNC is NOT defined:
master: (re)start mu4e.
before: Same as in master.

3) When mu4e is not running and FUNC is defined:
master: same as 1), and update contexts and requests.
before: same as master but always update contexts and requests.

4) When mu4e is not running and FUNC is NOT defined:
master: same as 2), and update contexts and requests.
before: same as master but always update contexts and requests.


> It might still be useful as optimization is having
> mu4e-index-updated-hook (and the :update / :remove messages) mark
> things as 'dirty'.. and skip the refresh when we're not dirty. But, at
> least for me, it works fine as it is.

You mean update counters when deleting messages?
This should be done in 4) (see below) by this PR.

So to resume, the requirements are:

1) Start mu4e with fresh info: ok.
2) Update main buffer with "g": ok but is should be revert-buffer that
do it.
3) Update main buffer after updating (after "U" finish or M-x
mu4e-update-index)
4) Update main buffer after reading unread mails and coming back from
mu4e-headers.


--
Thierry

Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

djcb commented:
I did a few more changes, I think they should address 1..4 (well, (2) works but does not use revert-buffer yet). I think the other things are working.

There's indeed a bit room for cleaning up the start-up process... I'll think a bit about that. Anyway, thank your!

thierryvolpiatto commented:
Dirk-Jan C. Binnema <notifications@github.com> writes:

> I did a few more changes, I think they should address 1..4

4) is not working yet.

Thanks.

-- 
Thierry

Get my Gnupg key:
gpg --keyserver pgp.mit.edu --recv-keys 59F29997

thierryvolpiatto commented:
I am now always using https://github.com/thierryvolpiatto/mu/tree/update_bmks_status
which IMO DTRT about revert-buffer and also works as expected, I am not sure using the
`mu4e-index-updated-hook` is the good approach specially in the mu4e source code for a specific problem.  Feel free to look at my branch.
Thanks.
